include/range/v3/view/chunk.hpp

```diff
@@ -14,114 +14,360 @@
+
             using offset_t =
-                meta::if_<
-                    BidirectionalRange<Rng>,
+                meta::if_c<
+                    BidirectionalRange<Rng>() || detail::CanSizedSentinel<Rng>(),
                     range_difference_type_t<Rng>,
                     constant<range_difference_type_t<Rng>, 0>>;
-            range_difference_type_t<Rng> n_;
-            friend range_access;
-            struct adaptor;
+
+            range_difference_type_t<Rng> n_ = 0;
+
+            struct adaptor
+              : adaptor_base, private box<offset_t>
+            {
+            private:
+                range_difference_type_t<Rng> n_;
+                sentinel_t<Rng> end_;
+                RANGES_CXX14_CONSTEXPR
+                offset_t const &offset() const
+                {
+                    offset_t const &result = this->box<offset_t>::get();
+                    RANGES_EXPECT(0 <= result && result < n_);
+                    return result;
+                }
+                RANGES_CXX14_CONSTEXPR
+                offset_t &offset()
+                {
+                    return const_cast<offset_t &>(const_cast<adaptor const &>(*this).offset());
+                }
+            public:
+                adaptor() = default;
+                constexpr adaptor(range_difference_type_t<Rng> n, sentinel_t<Rng> end)
+                  : box<offset_t>{0}, n_((RANGES_EXPECT(0 < n), n)), end_(end)
+                {}
+                RANGES_CXX14_CONSTEXPR
+                auto read(iterator_t<Rng> it) const ->
+                    decltype(view::take(make_iterator_range(std::move(it), end_), n_))
+                {
+                    RANGES_EXPECT(it != end_);
+                    RANGES_EXPECT(0 == offset());
+                    return view::take(make_iterator_range(std::move(it), end_), n_);
+                }
+                RANGES_CXX14_CONSTEXPR
+                void next(iterator_t<Rng> &it)
+                {
+                    RANGES_EXPECT(it != end_);
+                    RANGES_EXPECT(0 == offset());
+                    offset() = ranges::advance(it, n_, end_);
+                }
+                CONCEPT_REQUIRES(BidirectionalRange<Rng>())
+                RANGES_CXX14_CONSTEXPR
+                void prev(iterator_t<Rng> &it)
+                {
+                    ranges::advance(it, -n_ + offset());
+                    offset() = 0;
+                }
+                CONCEPT_REQUIRES(detail::CanSizedSentinel<Rng>())
+                RANGES_CXX14_CONSTEXPR
+                range_difference_type_t<Rng> distance_to(iterator_t<Rng> const &here,
+                    iterator_t<Rng> const &there, adaptor const &that) const
+                {
+                    auto const delta = (there - here) + (that.offset() - offset());
+                    // This can fail for cyclic base ranges when the chunk size does not divide the
+                    // cycle length. Such iterator pairs are NOT in the domain of -.
```

Comment:
`SizedSentinel` specifies that the difference must be well-defined when it is representable by the difference type; this view is non-conforming. Luckily it explodes the program instead of allowing non-conformance to be observed. I'm not terribly happy with that as a long-term solution, but I do think it is categorically better than the prior garbage-returning behavior.

It feels like overkill to remove difference support from `chunk_view` altogether; solving this permanently needs more thought.